File: apps/web/src/lib/api/taxa_contratada.ts
import { apiClient } from './client';

export interface TaxaContratadaCreate {
  bandeira: string;
  modalidade: string;
  taxa_contratada: number;
  vigencia_inicio: string; // ISO date
  vigencia_fim?: string | null;
  observacao?: string | null;
}

export interface TaxaContratadaResponse {
  id: number;
  cliente_id: number;
  bandeira: string;
  modalidade: string;
  taxa_contratada: number;
  vigencia_inicio: string;
  vigencia_fim: string | null;
  observacao: string | null;
  created_at: string;
}

export interface DesvioTaxa {
  bandeira: string;
  modalidade: string;
  taxa_contratada: number;
  taxa_media_cobrada: number;
  desvio_percentual: number;
  valor_total_transacoes: number;
  valor_excesso_estimado: number;
  status: 'ok' | 'atencao' | 'abusivo';
  quantidade_transacoes: number;
}

export interface ComparacaoResponse {
  cliente_id: number;
  processamento_id: string;
  desvios: DesvioTaxa[];
  valor_excesso_total: number;
}

export const taxaContratadaApi = {
  listar: (clienteId: number, vigente?: boolean): Promise<TaxaContratadaResponse[]> =>
    apiClient
      .get(`/clientes/${clienteId}/taxas-contratadas`, { params: vigente !== undefined ? { vigente } : {} })
      .then((r) => r.data),

  criar: (clienteId: number, data: TaxaContratadaCreate): Promise<TaxaContratadaResponse> =>
    apiClient.post(`/clientes/${clienteId}/taxas-contratadas`, data).then((r) => r.data),

  atualizar: (clienteId: number, id: number, data: Partial<TaxaContratadaCreate>): Promise<TaxaContratadaResponse> =>
    apiClient.put(`/clientes/${clienteId}/taxas-contratadas/${id}`, data).then((r) => r.data),

  remover: (clienteId: number, id: number): Promise<void> =>
    apiClient.delete(`/clientes/${clienteId}/taxas-contratadas/${id}`).then(() => undefined),

  comparacao: (clienteId: number, processamentoId: string): Promise<ComparacaoResponse> =>
    apiClient
      .get(`/clientes/${clienteId}/taxas-contratadas/comparacao`, { params: { processamento_id: processamentoId } })
      .then((r) => r.data),
};

File: apps/api/app/schemas/taxa_contratada.py
from datetime import date, datetime
from typing import List, Literal, Optional

from pydantic import BaseModel


class TaxaContratadaCreate(BaseModel):
    bandeira: str
    modalidade: str
    taxa_contratada: float
    vigencia_inicio: date
    vigencia_fim: Optional[date] = None
    observacao: Optional[str] = None


class TaxaContratadaUpdate(BaseModel):
    bandeira: Optional[str] = None
    modalidade: Optional[str] = None
    taxa_contratada: Optional[float] = None
    vigencia_inicio: Optional[date] = None
    vigencia_fim: Optional[date] = None
    observacao: Optional[str] = None


class TaxaContratadaResponse(BaseModel):
    id: int
    cliente_id: int
    bandeira: str
    modalidade: str
    taxa_contratada: float
    vigencia_inicio: date
    vigencia_fim: Optional[date] = None
    observacao: Optional[str] = None
    created_at: datetime

    model_config = {"from_attributes": True}


class DesvioTaxa(BaseModel):
    bandeira: str
    modalidade: str
    taxa_contratada: float
    taxa_media_cobrada: float
    desvio_percentual: float
    valor_total_transacoes: float
    valor_excesso_estimado: float
    status: Literal["ok", "atencao", "abusivo"]
    quantidade_transacoes: int


class ComparacaoResponse(BaseModel):
    cliente_id: int
    processamento_id: str
    desvios: List[DesvioTaxa]
    valor_excesso_total: float


class HistoricoDesvioItem(BaseModel):
    processamento_id: str
    data_processamento: Optional[datetime]
    desvios: List[DesvioTaxa]
    valor_excesso_total: float


class HistoricoDesviosResponse(BaseModel):
    cliente_id: int
    historico: List[HistoricoDesvioItem]
